### PR TITLE
1. Prepare for futurize stage 1 and clean up repository

### PR DESCRIPTION
**Pull request by amakhaba**
This PR does the following ([Prepare for Futurize stage 1](https://github.com/ska-sa/katdeveloper/wiki/Investigate-a-way-to-improve-Py3-migration-feedback-loop-on-tox#3-add-the-migration-tox-and-update-jenkinsfile-files)):

- Added the migration `tox.ini` file
- Updated the `Jenkinsfile` and `setup.py`
- Repository cleanup (grammar, imports and whitespaces related fixes)

*Definition of Done Checklist*

- [x] Unit tested (coded, passed, included)?
- [x] Requested at least 2 reviewers?

JIRA: [MT-738](https://skaafrica.atlassian.net/browse/MT-738)

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -36,22 +36,56 @@ pipeline {
...
+                    }
+                }
             }
+
             post {
                 always {
-                    junit 'nosetests.xml'
-                    cobertura coberturaReportFile: 'coverage.xml'
+                    junit 'nosetests_*.xml'
+                    cobertura (
```
Comment (mmphego) [inline Jenkinsfile]: This stage is going to fail due to the strictness you are imposing on your coverage report.

File: Jenkinsfile [hunk 2]
```diff
@@ -36,22 +36,56 @@ pipeline {
...
+                        failNoReports: true,
+                        failUnhealthy: true,
+                        failUnstable: true,
+                        autoUpdateHealth: true,
+                        autoUpdateStability: true,
+                        zoomCoverageChart: true,
+                        lineCoverageTargets: '80, 80, 80',
+                        conditionalCoverageTargets: '80, 80, 80',
+                        classCoverageTargets: '80, 80, 80',
+                        fileCoverageTargets: '80, 80, 80',
```
Comment (mmphego) [inline Jenkinsfile]: Please comment these lines out as this is too strict.

File: setup.py [hunk 3]
```diff
@@ -24,15 +24,6 @@
...
           "jsonschema"],
-      tests_require=[
-          'katcp',
-          'numpy',
-          'nose_xunitmp'],
-      extras_require={
-          'docs': ["sphinx-pypi-upload",
-                   "numpydoc",
-                   "Sphinx"],
-          },
```
Comment (mmphego) [inline setup.py]: Why was this removed?

File: setup.py [hunk 4]
```diff
@@ -24,15 +24,6 @@
           "PyTango>=9.2.2",
           "numpy",
           "jsonschema"],
-      tests_require=[
```
Comment (lanceWilliams) [inline setup.py]: How come is the 'tests_require' removed? Is it unneeded?

File: Jenkinsfile (hunk 1 above)
Comment (amakhaba) [reply Jenkinsfile]: Commented out the strictness in the `Jenkinsfile`

File: setup.py (hunk 3 above)
Comment (amakhaba) [reply setup.py]: This is covered in the `tox.ini` file under the dependencies

File: setup.py (hunk 4 above)
Comment (amakhaba) [reply setup.py]: See: https://github.com/ska-sa/tango-simlib/pull/105#discussion_r353102050

Comment (lanceWilliams) [reply setup.py]: Don't we still need it if we are just running nose tests in the module?

File: Jenkinsfile (hunk 2 above)
Comment (amakhaba) [reply Jenkinsfile]: Done thanks

File: setup.py (hunk 3 above)
Comment (amakhaba) [reply setup.py]: Added these to the tox file that was later added after your comment thanks.

File: setup.py (hunk 4 above)
Comment (amakhaba) [reply setup.py]: Ideally what was listed here is listed under dependencies in the tox file. and tox should be able to invoke nose tests directly.

File: Jenkinsfile [hunk 5]
```diff
@@ -41,17 +48,48 @@ pipeline {
...
+            environment {
+                test_flags = "${KATPACKAGE}"
             }
+
+             parallel {
+                stage ('py27') {
+                    steps {
+                        echo "Running nosetests on Python 2.7"
+                        sh 'python2 -m pip install . -U'
+                        sh 'python2 -m pip install nose_xunitmp katcp'
```
Comment (ajoubertza) [inline Jenkinsfile]: katcp should be a test requirement rather than installing it manually here

File: setup.py (hunk 4 above)
Comment (ajoubertza) [reply setup.py]: We need to re-instate this since we are not going with tox.

File: setup.py (hunk 3 above)
Comment (ajoubertza) [reply setup.py]: Are you proposing that tox is used for document generation, but not for testing?

File: Jenkinsfile [hunk 6]
```diff
@@ -23,16 +23,23 @@ pipeline {
             }
         }
 
-        stage ('Static analysis') {
-            steps {
-                sh "pylint ./${KATPACKAGE} --output-format=parseable --exit-zero > pylint.out"
-                //sh "lint_diff.sh -r ${KATPACKAGE}"
-            }
+        // stage ('Static analysis') {
```
[the remark this answers is not in the document]
Comment (amakhaba) [reply Jenkinsfile]: Judging from @mmphego commit message i guess he wanted a faster and efficient feedback whilst we were busy with the tox failure debugging

File: setup.py (hunk 3 above)
Comment (amakhaba) [reply setup.py]: As per agreement with @mmphego and @tockards we first wanted your take on making an exception for `tango-simlib` not to use tox, and after that we will refactor everything to be what they need to be.

File: Jenkinsfile (hunk 6 above)
[the remark this answers is not in the document]
Comment (ajoubertza) [reply Jenkinsfile]: OK, but please add a "TODO" comment so that we can see that.

File: setup.py (hunk 3 above)
Comment (ajoubertza) [reply setup.py]: Yes, it is OK for to leave out tox for tango-simlib.

**Comment by mmphego** [comment]
Looks good, but I think at this current moment we do not need tox. We can continue without it as the tests are run via python nosetests and I do not think we should only keep tox for docs generation (Wouldn't make sense).

I suppose, if we have enough time towards the end of the sprint then we can revisit the use of tox and how we can get tests to run but till then please remove the tox.ini file.

@ajoubertza and @tockards what do you think?

File: setup.py (hunk 3 above)
Comment (mmphego) [reply setup.py]: Please restore these options on setup.py

**Comment by ajoubertza** [comment]
> I suppose, if we have enough time towards the end of the sprint then we can revisit the use of tox and how we can get tests to run but till then please remove the tox.ini file.

No tox for this repo is fine.  I don't mind if it is removed now or in a later PR.  When removed, then restore the `extras_require` as well.

File: tox.ini [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                xargs futurize --stage2 --write --all-imports --nobackups"
+    bash -c "grep --exclude=tox.ini -rl 'from __future__ import unicode_literals' | \
+                xargs sed -e 's/from __future__ import unicode_literals//g' -i"
+    bash -c "grep --exclude=tox.ini -rl 'from builtins import str' | \
+                xargs sed -e 's/from builtins import str//g' -i"
+    bash -c "grep --exclude=tox.ini -rl 'from builtins import *' | \
+                xargs sed -e 's/from builtins import \*//g' -i"
+
+whitelist_externals =
+    {[testenv]whitelist_externals}
```
Comment (mmphego) [inline tox.ini]: Remove `tox.ini` from repo and restore changes made on `setup.py`.

File: tango_simlib/model.py [hunk 8]
```diff
@@ -168,7 +168,7 @@ class PopulateModelQuantities(object):
     """Used to populate/update model quantities.
 
     Populates the model quantities using the data from the TANGO device information
-    captured in the json file / POGO generated xmi / FANDANGO generated fgo file.
+    captured in the json file / POGO generated xmi / FANDANGO generated to file.
```
Comment (ajoubertza) [inline tango_simlib/model.py]: This was not a typo - `fgo` is the file extension

**Comment by mmphego** [comment]
@ajoubertza Additional cleanup done, this time to the whole repository.

PR ready to be merged.
